microproxy/proxy.py

```diff
@@ -14,53 +14,49 @@
 
 
 class LayerManager(object):
-    def __init__(self, src_stream, config):
-        self.src_stream = src_stream
+    def __init__(self, config):
         self.config = config
 
     @gen.coroutine
-    def start_layer(self):
-        mode = self.config["mode"]
-        if mode == "socks":
-            layer_constructor = SocksLayer
-        elif mode == "transparent":
-            layer_constructor = TransparentLayer
-        else:
-            logger.error("Unsupport proxy mode : {0}".format(mode))
-            raise ValueError
-
-        current_context = Context(src_stream=self.src_stream,
+    def run_layers(self, src_stream):
+        current_context = Context(src_stream=src_stream,
                                   config=self.config)
-        while True:
+        current_layer = self.get_first_layer(current_context)
+
+        while current_layer:
             try:
-                current_layer = layer_constructor(current_context)
                 logger.debug("Enter {0} Layer".format(current_layer))
-                current_context = yield current_layer.process_and_return_context()
+                current_context = yield gen.maybe_future(current_layer.process_and_return_context())
```

Comment:
Unfortunately, according to tornado documentation, the ```maybe_future``` has been deprecated.
http://www.tornadoweb.org/en/stable/gen.html#tornado.gen.maybe_future

```
Deprecated since version 4.3: This function only handles Futures, not other yieldable objects. Instead of maybe_future, check for the non-future result types you expect (often just None), and yield anything unknown.
```
We may need to find another way to handle this part